src/txrequest.h

```diff
@@ -0,0 +1,342 @@
+ *
+ * The following operations are supported on this data structure:
+ *
+ * - ReceivedInv(txid, peer) adds a new CANDIDATE entry, unless one already exists for that (txid, peer) combination
+ *   (whether it's CANDIDATE, REQUESTED, or COMPLETED).
+ *
+ * - DeletedPeer(peer) deletes all entries for a given peer. It should be called when a peer goes offline.
+ *
+ * - AlreadyHaveTx(txid) deletes all entries for a given txid. It should be called when a transaction is succesfully
+ *   added to the mempool, seen in a block, or for whatever reason we no longer care about it.
+ *
+ * - ReceivedResponse(peer, txid) converts any CANDIDATE or REQUESTED entry to a COMPLETED one. It should be called
+ *   whenever a transaction or NOTFOUND was received from a peer. When the transaction is acceptable, AlreadyHaveTx
+ *   should be called instead of (or in addition to) this operation.
+
+ * - GetRequestable(peer) does the following:
+ *   - Convert all REQUESTED entries (for all txids/peers) with (exptime <= now) to COMPLETED entries.
+ *   - Requestable txids are selected: txids which have no REQUESTED entry, and for which this peer has the
+ *     lowest-priority CANDIDATE entry among all CANDIDATEs with (reqtime <= now) at the time of the GetRequestable
+ *     call. As the high bit of the priority is whether or not the peer is inbound, this will always prefer an
+ *     outbound peer as long as one is available.
+ *   - The selected txids are sorted in order of announcement and returned (even if multiple were added at the same
+ *     time, or even when the clock went backwards while they were being added).
+ *
+ * - RequestedTx(peer, txid) converts the CANDIDATE entry for the provided peer and txid into a REQUESTED one, with
+ *   exptime set to (now + timeout). It can ONLY be called immediately after GetRequestable was called (for the same
+ *   peer), with only AlreadyHaveTx and other RequestedTx calls (both for other txids) in between. Any other
+ *   non-const operation removes the ability to call RequestedTx.
+ */
+class TxRequestTracker {
+    //! Configuration option: delay before downloading from inbound connection.
+    const std::chrono::microseconds m_inbound_delay;
+    //! Configuration option: timeout after requesting a download.
+    const std::chrono::microseconds m_timeout;
+
+    //! The various states a (txid,node) pair can be in.
+    //! Note that CANDIDATE is split up into 3 substates (NEW, BEST, OTHER), allowing more efficient implementation.
+    enum class State : uint8_t {
+        //! A CANDIDATE entry whose reqtime is in the future.
+        CANDIDATE_NEW,
+        //! The best CANDIDATE for a given txid (lowest priority among all outbound if any; among inbound otherwise),
+        //! assuming there is no REQUESTED entry already.
+        CANDIDATE_BEST,
+        //! A REQUESTED entry.
+        REQUESTED,
+        //! A CANDIDATE entry that's not CANDIDATE_NEW or CANDIDATE_BEST.
+        CANDIDATE_OTHER,
```

Comment:
nit: sort this above `REQUESTED` so it's next to the other `CANDIDATE_` entries.

Reply:
That doesn't work directly, as EntryTxid's sort order relies on the ordering of these fields. It could use its own enum with the necessary order, that ExtractTxid maps to, but that feels like overkill.

Reply:
Maybe worth adding a comment in the enum definition that the ordering is relied upon by the code that uses the ByTxid indexes. Reordering in ExtractTxid seems like overkill to me too fwiw.

Reply:
Done.

Reply:
Agree that a comment is sufficient.